docs/content/guides/developer/app-examples/trustless-swap.mdx

```diff
@@ -1,27 +1,1957 @@
-1. [Frontend](./trustless-swap/frontend.mdx): Enables users to list objects for sale and to accept trades.
+The guide also shows how to build an app that:
 
-{@include: ../../../snippets/app-examples-swap-source.mdx}
+- **Is trustless:** Users do not have to trust (or pay) any third parties; the chain manages the swap.
+- **Avoids rug-pulls:** Guarantees that the object a user wants to trade for isn't tampered with after initiating the trade.
+- **Preserves liveness:** Users are able to pull out of the trade and reclaim their object at any time, in case the other party stops responding.
 
-## Prerequisites
+## What you need
+
+{@include: ../../../snippets/app-examples-swap-source.mdx}
 
 Before getting started, make sure you have:
 
 - [Installed the latest version of Sui](../getting-started/sui-install.mdx).
-- [Configured a valid network environment](../../../references/cli/client.mdx#set-current-environment), as you will be deploying the module on Testnet.
+- [Configured a valid network environment](../../../references/cli/client.mdx#set-current-environment), as the guide has you deploy the module on Testnet.
+- [Acquired Devnet or Testnet](../getting-started/get-coins.mdx) tokens for development purposes.
+- Read the basics of [shared versus owned objects](../sui-101/shared-owned.mdx).
+
+## Smart contracts {#smart-contracts}
+
+In this part of the guide, you write the Move contracts that perform the trustless swaps. The first step is to [set up a Move package](../first-app/write-package.mdx) for storing your Move modules.
+
+### Locked and Key
+
+After a trade is initiated, you don't want the trading party to modify the object they agreed to trade. Imagine you're trading in-game items and you agree to trade a weapon with all its attachments, and its owner strips all its attachments just before the trade.
+
+In a traditional trade, a third party typically holds the items in escrow to make sure they are not tampered with before the trade completes. This requires either trust that the third party won't tamper with it themselves, paying the third party to ensure that doesn't happen, or both.
+
+In a trustless swap, however, you can use the safety properties of Move to force an item's owner to prove that they have not tampered with the version of the object that you agreed to trade, without involving anyone else.
+
+This is done by requiring that an object that is available for trading is **locked** with a **single-use key**, and asking the owner to supply the key when finalizing the trade.
+
+To tamper with the object would require unlocking it, which consumes the key. Consequently, there would no longer be a key to finish the trade.
+
+{@inject: examples/trading/contracts/escrow/sources/lock.move#struct=Locked,Key noComments}
+
+- The `Locked<T>` type stores the locked `obj`, the `ID` of the key that unlocks it, and its own `id`.
```

Comment:
Updating to reflect the fact that the snippet now shows the dynamic field version.

```suggestion
- The `Locked<T>` type stores the `ID` of the key that unlocks it, and its own `id`. The object being locked will be added as a dynamic object field, so that it is still readable at its own ID off-chain.
```

Can we also add a link on "dynamic object field" or link to the appropriate concept?

I also left this comment on the previous PR:

> The ideal case is that we explain this progression: start with the code that stores the obj inline, and then when we need to display the object in the front-end, we explain the requirement, and the guide runs someone through updating the Locked type to make that work at that point. But let's add that as a future improvement?

Can we track that future improvement somewhere?